source/base/cuda.cu

```diff
@@ -45,18 +45,6 @@ namespace Utilities
 
     Handle::~Handle()
     {
-      dealii::GrowingVectorMemory<
-        LinearAlgebra::CUDAWrappers::Vector<float>>::release_unused_memory();
-      dealii::GrowingVectorMemory<
```

Comment:
Hm, where is the `Handle` object created? If the handle is a local variable in `main`? If so, then the handle is created before the pool (which is now a function-local `static` variable), but it's destroyed before the pool is destroyed -- namely, just before the end of `main()`, whereas all `static` variables are destroyed *at* the end of main.

Reply:
The `Handle` exists because without it, the pool is destroyed after the CUDA environment is destroyed. This is a problem because we are trying to call `cudaFree` but the CUDA environment does not exist anymore. If the new implementation destroys the pool before the CUDA environment is teared down, then there is no point in releasing the memory in the `Handle`.

Reply:
> whereas all `static` variables are destroyed _at_ the end of main.

Thanks for correcting me being unprecise/misled here.

Reply:
I think I didn't express my thoughts clearly.

What I meant to say is that removing these lines relies on the fact that the `Handle` is *also* a `static` variable. Then the ordering is correct: Since presumably the `Handle` is created before the pool is accessed the first time, the vectors are freed before the `Handle` is destroyed.

But this is not the case if the `Handle` is simply a regular local variable in `main()` for which the ordering guarantee does not hold because the ordering is only among all of those variables that are `static`.

Reply:
I think that there is a misunderstanding here. The `Handle` does not have something similar to `MPI_Init`. The `CUDA` environment is initialized after all(?) the global variable but before going in the main. So as long as the static variables are destroyed before the `CUDA` environment this works. The `Handle` initializes `CUDA` libraries but not the `CUDA` environment which is what matters here.

Reply:
In fact, there is no `static` `Handle` anymore. In all the tests that use a `CUDA::Handle`, it is only a local variable.

Reply:
Do you have an example where we create such a Handle?

Reply:
To give an example from the testsuite:
```
int
main()
{
  initlog();
  deallog.depth_console(0);

  init_cuda();

  Utilities::CUDA::Handle cuda_handle;
  test(cuda_handle);

  deallog << "OK" << std::endl;

  return 0;
}
```
Here, at the `return 0;` statement, the local `cuda_handle` variable is destroyed which tears down the handle. But the destructor of all `static` variables only runs *after* that, i.e., when the handle is already gone. Is this what you want?

Reply:
> Here, at the return 0; statement, the local cuda_handle variable is destroyed which tears down the handle. But the destructor of all static variables only runs after that, i.e., when the handle is already gone. Is this what you want?

That's fine.

We used to hit this [problem](https://stackoverflow.com/questions/43235548/how-to-handle-cudafree-on-globally-instantiated-variables). The problem is that the driver is being unloaded at the end of the program but before the global variables are cleaned. We are not talking about some "regular" libraries but about the `CUDA` drivers which are unloaded when they are not necessary anymore. The loading/unloading of the driver is different than for other library. BTW what I said before was wrong, the driver is loaded the first time we need `CUDA`.